segmentstore/server/src/main/java/io/pravega/segmentstore/server/security/TLSConfigChangeHandler.java

```diff
@@ -0,0 +1,43 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.server.security;
+
+import io.netty.handler.ssl.SslContext;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+@RequiredArgsConstructor
+@Slf4j
+public class TLSConfigChangeHandler {
+
+    /**
+     * A counter representing the number of times this object has been asked to
+     * consume an event.
+     */
+    private final AtomicInteger numOfConfigChangesSinceStart = new AtomicInteger(0);
+
+    private @NonNull final AtomicReference<SslContext> sslContext;
+    private @NonNull final String pathToCertificateFile;
+    private @NonNull final String pathToKeyFile;
+
+    public void handleTlsConfigChange() {
+        log.info("Current reload count = {}", numOfConfigChangesSinceStart.incrementAndGet());
+        sslContext.set(TLSHelper.newServerSslContext(pathToCertificateFile, pathToKeyFile));
```

Comment:
I'm not sure I understand this flow. This line sets a new context through `TLSHelper`, but I don't understand how this is actually loaded. Setting the atomic reference shouldn't cause it to be automatically loaded by Netty.

I'm also missing where the channels are being closed upon reload (that's what the description says).  I'm particularly concerned about it because dropping all connections at once will cause all clients to try to reconnect together, causing a herd effect for a large number of clients. I'm wondering if instead of closing all channels, we can:

- Not close them at all and simply create new channels with the new context
- Close channels gradually

Reply:
Sorry, the description still had a statement from the approach followed initially, which is now replaced with a different approach (based on the discussions we have had above): 

We now share the same SSL Context across all channels (like we did earlier). Upon detection that the certificate has changed, the callback creates a new SSL Context and assigns it to the shared variable. So, existing connections established using old SSL Context continue to function, while new ones use the new SSL Context built from the new certificate (and the key).